File: articles/design-thinking-insight-generation/go/insight_value_engine.go
package main

import (
	"encoding/csv"
	"fmt"
	"log"
	"os"
	"sort"
	"strconv"
)

type Insight struct {
	Name             string
	PatternSupport   float64
	ExplanatoryDepth float64
	OpportunityValue float64
	InterpretiveRisk float64
	Value            float64
}

func parseFloat(value string) float64 {
	parsed, err := strconv.ParseFloat(value, 64)
	if err != nil {
		log.Fatalf("unable to parse numeric value %q: %v", value, err)
	}
	return parsed
}

func insightValue(insight Insight) float64 {
	return 0.30*insight.PatternSupport +
		0.30*insight.ExplanatoryDepth +
		0.25*insight.OpportunityValue -
		0.15*insight.InterpretiveRisk
}

func loadInsights(path string) []Insight {
	file, err := os.Open(path)
	if err != nil {
		log.Fatalf("unable to open input file: %v", err)
	}
	defer file.Close()

	reader := csv.NewReader(file)
	rows, err := reader.ReadAll()
	if err != nil {
		log.Fatalf("unable to read CSV: %v", err)
	}

	var insights []Insight

	for i, row := range rows {
		if i == 0 {
			continue
		}
		if len(row) < 5 {
			log.Fatalf("malformed row: %v", row)
		}

		insight := Insight{
			Name:             row[0],
			PatternSupport:   parseFloat(row[1]),
			ExplanatoryDepth: parseFloat(row[2]),
			OpportunityValue: parseFloat(row[3]),
			InterpretiveRisk: parseFloat(row[4]),
		}
		insight.Value = insightValue(insight)
		insights = append(insights, insight)
	}

	return insights
}

func main() {
	path := "../data/raw/candidate_insights_raw.csv"
	if len(os.Args) > 1 {
		path = os.Args[1]
	}

	insights := loadInsights(path)

	sort.Slice(insights, func(i, j int) bool {
		return insights[i].Value > insights[j].Value
	})

	fmt.Println("rank,insight,insight_value")
	for i, insight := range insights {
		fmt.Printf("%d,%s,%.6f\n", i+1, insight.Name, insight.Value)
	}
}

File: articles/design-thinking-insight-generation/c/insight_value_engine.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_INSIGHTS 128
#define MAX_LINE 512
#define MAX_NAME 180

typedef struct {
    char name[MAX_NAME];
    double pattern_support;
    double explanatory_depth;
    double opportunity_value;
    double interpretive_risk;
    double value;
} Insight;

double insight_value(const Insight *insight) {
    const double wp = 0.30;
    const double we = 0.30;
    const double wo = 0.25;
    const double wr = 0.15;

    return wp * insight->pattern_support
         + we * insight->explanatory_depth
         + wo * insight->opportunity_value
         - wr * insight->interpretive_risk;
}

int compare_desc(const void *a, const void *b) {
    const Insight *ia = (const Insight *)a;
    const Insight *ib = (const Insight *)b;

    if (ia->value < ib->value) return 1;
    if (ia->value > ib->value) return -1;
    return 0;
}

int main(int argc, char **argv) {
    const char *path = argc > 1 ? argv[1] : "../data/raw/candidate_insights_raw.csv";
    FILE *file = fopen(path, "r");

    if (!file) {
        fprintf(stderr, "ERROR: Unable to open input file: %s\n", path);
        return 1;
    }

    Insight insights[MAX_INSIGHTS];
    char line[MAX_LINE];
    int count = 0;

    fgets(line, sizeof(line), file); /* header */

    while (fgets(line, sizeof(line), file) && count < MAX_INSIGHTS) {
        char *token = strtok(line, ",");
        if (!token) continue;

        strncpy(insights[count].name, token, MAX_NAME - 1);
        insights[count].name[MAX_NAME - 1] = '\0';

        token = strtok(NULL, ",");
        insights[count].pattern_support = token ? atof(token) : 0.0;

        token = strtok(NULL, ",");
        insights[count].explanatory_depth = token ? atof(token) : 0.0;

        token = strtok(NULL, ",");
        insights[count].opportunity_value = token ? atof(token) : 0.0;

        token = strtok(NULL, ",");
        insights[count].interpretive_risk = token ? atof(token) : 0.0;

        insights[count].value = insight_value(&insights[count]);
        count++;
    }

    fclose(file);

    qsort(insights, count, sizeof(Insight), compare_desc);

    printf("rank,insight,insight_value\n");
    for (int i = 0; i < count; i++) {
        printf("%d,%s,%.6f\n", i + 1, insights[i].name, insights[i].value);
    }

    return 0;
}
